File: oneflow/core/register/blob_desc.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/register/blob_desc.h"

namespace oneflow {

bool CompareLbiBlobDescPair(const LbiBlobDescPair& lhs, const LbiBlobDescPair& rhs) {
  return lhs.lbi() < rhs.lbi();
}

BlobDesc::BlobDesc(const Shape& shape, DataType dtype, bool is_dynamic)
    : shape_(SymbolOf(shape)),
      stride_(SymbolOf(Stride(shape))),
      data_type_(dtype),
      is_dynamic_(is_dynamic) {}
BlobDesc::BlobDesc(const Shape& shape, const Stride& stride, DataType dtype, bool is_dynamic)
    : shape_(SymbolOf(shape)),
      stride_(SymbolOf(stride)),
      data_type_(dtype),
      is_dynamic_(is_dynamic) {}
BlobDesc::BlobDesc(Symbol<Shape> shape, Symbol<Stride> stride, DataType dtype, bool is_dynamic)
    : shape_(shape), stride_(stride), data_type_(dtype), is_dynamic_(is_dynamic) {}
BlobDesc::BlobDesc(const Shape& shape, DataType dtype)
    : BlobDesc(shape, Stride(shape), dtype, false) {}
BlobDesc::BlobDesc(const Shape& shape, const Stride& stride, DataType dtype)
    : BlobDesc(shape, stride, dtype, false) {}
BlobDesc::BlobDesc(Symbol<Shape> shape, Symbol<Stride> stride, DataType dtype)
    : BlobDesc(shape, stride, dtype, false) {}
BlobDesc::BlobDesc(DataType dtype) : BlobDesc(Shape(), Stride(), dtype, false) {}

BlobDesc::BlobDesc(const BlobDescProto& proto) {
  shape_ = SymbolOf(Shape(proto.shape()));
  stride_ = SymbolOf(Stride(proto.stride()));
  data_type_ = proto.data_type();
  is_dynamic_ = proto.is_dynamic();
}

BlobDesc::BlobDesc(const BlobDesc& other)
    : shape_(other.shape_),
      stride_(other.stride_),
      data_type_(other.data_type()),
      is_dynamic_(other.is_dynamic()) {}

void BlobDesc::ToProto(BlobDescProto* proto) const {
  shape().ToProto(proto->mutable_shape());
  stride().ToProto(proto->mutable_stride());
  proto->set_data_type(data_type_);
  proto->set_is_dynamic(is_dynamic_);
}

BlobDesc& BlobDesc::operator=(const BlobDesc& rhs) {
  this->CopyFrom(rhs);
  return *this;
}

void BlobDesc::CopyFrom(const BlobDesc& other) {
  set_shape(other.shape());
  set_stride(other.stride());
  set_data_type(other.data_type());
  set_is_dynamic(other.is_dynamic());
}

void BlobDesc::set_is_dynamic(bool is_dynamic) { is_dynamic_ = is_dynamic; }

bool BlobDesc::operator==(const BlobDesc& rhs) const {
  return (shape() == rhs.shape()) && (stride() == rhs.stride()) && (data_type() == rhs.data_type())
         && (is_dynamic() == rhs.is_dynamic());
}

size_t BlobDesc::ByteSizeOfBlobHeader() const {
  return shape().is_initialized() ? shape().NumAxes() * sizeof(int64_t) : 0;
}

size_t BlobDesc::AlignedByteSizeOfBlobHeader() const {
  return shape().is_initialized()
             ? RoundUp(shape().NumAxes() * sizeof(int64_t), kBlobHeaderAlignSize)
             : RoundUp(0, kBlobHeaderAlignSize);
}

size_t BlobDesc::ByteSizeOfBlobBody() const {
  return shape().is_initialized() ? shape().elem_cnt() * GetSizeOfDataType(data_type()) : 0;
}

size_t BlobDesc::AlignedByteSizeOfBlobBody() const {
  return RoundUp(ByteSizeOfBlobBody(), kBlobBodyAlignSize);
}

size_t BlobDesc::AlignedTotalByteSize() const {
  return AlignedByteSizeOfBlobHeader() + AlignedByteSizeOfBlobBody();
}

}  // namespace oneflow

File: oneflow/python/framework/blob_desc.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from __future__ import absolute_import

import copy
import traceback

import oneflow.core.common.data_type_pb2 as data_type_util
import oneflow.python.framework.distribute as distribute_util
import oneflow.python.lib.core.traceinfo as traceinfo


class BlobDesc(object):
    def __init__(self, lbi, distribute=distribute_util.auto(), disable_boxing=None):
        self.lbi_ = lbi
        self.lbn_ = lbi.op_name + "/" + lbi.blob_name
        self.distribute_ = distribute
        self.disable_boxing_ = disable_boxing

    @property
    def lbi(self):
        return self.lbi_

    @property
    def logical_blob_name(self):
        return self.lbn_

    @property
    def op_name(self):
        return self.lbi_.op_name

    @property
    def blob_name(self):
        return self.lbi_.blob_name

    @property
    def shape(self):
        raise NotImplementedError

    @property
    def dtype(self):
        raise NotImplementedError

    @property
    def batch_axis(self):
        raise NotImplementedError

    def has_batch_axis(self):
        batch_axis = self.batch_axis
        ret = batch_axis is not None
        if ret:
            assert type(batch_axis) is int
        return ret

    @property
    def is_dynamic(self):
        raise NotImplementedError

    @property
    def disable_boxing(self):
        return self.disable_boxing_

    @property
    def is_tensor_list(self):
        raise NotImplementedError

    @property
    def parallel_conf(self):
        raise NotImplementedError

    def with_boxing_disabled(self, val=True):
        ret = self.Clone()
        ret.disable_boxing_ = val
        return ret

    def with_distribute(self, distribute):
        ret = self.Clone()
        ret.distribute_ = distribute
        return ret

    def with_split_distribute(self, axis):
        return self.with_distribute(distribute_util.split(axis))

    def with_broadcast_distribute(self):
        return self.with_distribute(distribute_util.broadcast())

    @property
    def distribute(self):
        distribute_util.assert_is_valid_distribute(self.distribute_)
        return self.distribute_

    @property
    def unique_name(self):
        return self.lbn_ + self._Distribute2Str() + self._DisableBoxing2Str()

    def Clone(self):
        return copy.deepcopy(self)

    def _Distribute2Str(self):
        if type(self.distribute_) is distribute_util.AutoDistribute:
            return ""
        elif type(self.distribute_) is distribute_util.SplitDistribute:
            return ":S" + str(self.distribute_.axis)
        elif type(self.distribute_) is distribute_util.BroadcastDistribute:
            return ":B"
        else:
            raise NotImplementedError

    def _DisableBoxing2Str(self):
        if self.disable_boxing_ is None:
            return ""
        if self.disable_boxing_ is False:
            return "|0"
        if self.disable_boxing_ is True:
            return "|1"
        raise NotImplementedError
